### PR TITLE
perf: use popcnt in tr_bitfield

### PR DESCRIPTION
**Pull request by goldsteinn**
Same as [PR #2938](https://github.com/transmission/transmission/issues/2938) but instead of relying on [libpopcnt](https://github.com/transmission/libpopcnt) just makes new header file [tr-popcnt.h](https://github.com/transmission/transmission/compare/main...goldsteinn:submittable-hdr?expand=1#diff-ef11ec5a5d3fd4a66efbaafcd1ff9933faa5db09d766ec5f5bac20556ed13131) that accomplishes the same logic (but better for smaller types such as the `uint8_t` currently used).

Some notes: Tested with GCC11 on ubuntu20.04

Tested `tr_popcnt` independently with GCC11/Clang12 on ubuntu20.04 but the MSVC  preprocessor logic is untested. 
All `count` implementations are tested, however.

All tests either pass or are skipped:

```
$> libtransmission-test
...
[----------] 5 tests from Bitfield
[ RUN      ] Bitfield.count
[       OK ] Bitfield.count (1156 ms)
[ RUN      ] Bitfield.ctorFromFlagArray
[       OK ] Bitfield.ctorFromFlagArray (0 ms)
[ RUN      ] Bitfield.setRaw
[       OK ] Bitfield.setRaw (0 ms)
[ RUN      ] Bitfield.bitfields
[       OK ] Bitfield.bitfields (0 ms)
[ RUN      ] Bitfield.hasAllNone
[       OK ] Bitfield.hasAllNone (0 ms)
[----------] 5 tests from Bitfield (1156 ms total)
...
[ RUN      ] CompletionTest.createPieceBitfield
[       OK ] CompletionTest.createPieceBitfield (2 ms)
...
[----------] Global test environment tear-down
[==========] 270 tests from 37 test suites ran. (26791 ms total)
[  PASSED  ] 252 tests.
[  SKIPPED ] 18 tests, listed below:
[  SKIPPED ] JSON/JSONTest.testElements/1
[  SKIPPED ] JSON/JSONTest.testElements/2
[  SKIPPED ] JSON/JSONTest.testElements/3
[  SKIPPED ] JSON/JSONTest.testUtf8/1
[  SKIPPED ] JSON/JSONTest.testUtf8/2
[  SKIPPED ] JSON/JSONTest.testUtf8/3
[  SKIPPED ] JSON/JSONTest.test1/1
[  SKIPPED ] JSON/JSONTest.test1/2
[  SKIPPED ] JSON/JSONTest.test1/3
[  SKIPPED ] JSON/JSONTest.test2/1
[  SKIPPED ] JSON/JSONTest.test2/2
[  SKIPPED ] JSON/JSONTest.test2/3
[  SKIPPED ] JSON/JSONTest.test3/1
[  SKIPPED ] JSON/JSONTest.test3/2
[  SKIPPED ] JSON/JSONTest.test3/3
[  SKIPPED ] JSON/JSONTest.unescape/1
[  SKIPPED ] JSON/JSONTest.unescape/2
[  SKIPPED ] JSON/JSONTest.unescape/3
```

### CODE REVIEW

File: libtransmission/tr-popcount.h [hunk 1]
```diff
@@ -0,0 +1,213 @@
+// This file Copyright © 2008-2022 Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
```
Comment (ckerr) [inline libtransmission/tr-popcount.h:4]: If this code borrows any pieces from libpopcnt, those pieces should retain their original copyright message

For the new wrapper pieces, remove the `2008-`

Comment (goldsteinn) [reply libtransmission/tr-popcount.h:4]: It doesn't borrow anything. Will fixup 2008-2022.

Comment (goldsteinn) [reply libtransmission/tr-popcount.h:4]: What is the expected size of the bitfield btw? If we are expecting large bitfields could make the code significantly faster. I made more original patch so there would be no regressions but even a simple unrolling of the loop to better hide the latency gets significant gain (although costs a bit for small sizes)

Comment (goldsteinn) [reply libtransmission/tr-popcount.h]: Fixed in V2.

**Comment by goldsteinn** [comment]
Just want to add. Comparing the `libpopcnt` version you where going to use vs. main:

```
bits, Time Main, Time Libpopcnt, Time Main / Time Libpopcnt
   1,     0.147,          0.183,                      0.803
   8,     0.145,          0.209,                      0.693
  15,     0.263,          0.210,                      1.252
  22,     0.256,          0.206,                      1.248
  29,     0.307,          0.196,                      1.564
  36,     0.361,          0.207,                      1.740
  43,     0.402,          0.234,                      1.715
  50,     0.458,          0.248,                      1.848
  57,     0.507,          0.257,                      1.976
  64,     0.506,          0.249,                      2.028
 320,     2.180,          0.285,                      7.653
 576,     4.053,          0.323,                     12.533
 832,     5.663,          0.357,                     15.847
1088,     7.316,          0.401,                     18.256
1344,     8.956,          0.448,                     19.994
1600,    10.620,          0.497,                     21.352
1856,    12.262,          0.551,                     22.260
2112,    13.927,          0.596,                     23.354
2368,    15.547,          0.667,                     23.322
2624,    17.174,          0.743,                     23.111
2880,    18.887,          0.839,                     22.523
3136,    20.608,          0.986,                     20.902
3392,    22.168,          1.014,                     21.858
3648,    23.836,          1.079,                     22.086
3904,    25.518,          1.128,                     22.627
4160,    27.329,          0.614,                     44.523
4416,    28.761,          0.648,                     44.372
4672,    30.508,          0.670,                     45.546
4928,    32.272,          0.693,                     46.585
5184,    33.956,          0.711,                     47.780
5440,    35.505,          0.727,                     48.835
5696,    37.093,          0.745,                     49.756
5952,    38.850,          0.763,                     50.933
6208,    40.472,          0.780,                     51.855
6464,    42.100,          0.798,                     52.727
6720,    43.661,          0.817,                     53.414
6976,    45.521,          0.837,                     54.391
7232,    47.375,          0.856,                     55.358
7488,    48.858,          0.874,                     55.930
7744,    50.523,          0.891,                     56.704
8000,    52.098,          0.911,                     57.166
8192,    53.388,          0.925,                     57.715
```

Not arguing to use `libpopcnt`. But if `bitfield` is expected to be large
there is significant benefit to implementing a more optimized loop.

**Comment by goldsteinn** [comment]
Test failure appears to be timeout of build itself on mac. Don't see how this is related to the change.

**Comment by ckerr** [comment]
> What is the expected size of the bitfield btw? If we are expecting large bitfields could make the code significantly faster. I made more original patch so there would be no regressions but even a simple unrolling of the loop to better hide the latency gets significant gain (although costs a bit for small sizes)

There is no real upper bound. Yes, these bitfields can potentially be very large when they are representing mega-torrents, so if you want to widen / unroll / etc. and have benchmarks to demonstrate the improvements I'd be happy  to consider for inclusion.

**Comment by ckerr** [comment]
The macOS CI failure is not a blocker for this PR but I'm going to leave it open for now since @goldsteinn has indicated there could be more commits

**Comment by goldsteinn** [comment]
> The macOS CI failure is not a blocker for this PR but I'm going to leave it open for now since @goldsteinn has indicated there could be more commits

Will be a bit before I have time to circle back to optimize the loop. Would prefer this go out first and can

**Comment by ckerr** [comment]
OK just a quick checkin -- @goldsteinn I saw a comment earlier (now deleted?) that you'd rather this PR go in as-is and then have another followup to improve speed. If that's still the case, lmk and I'll merge this one.

**Pull request merged**